File: algorithms-practice/graphs/number_of_islands.py
"""
Number of Islands (DFS/BFS on grid of '0'/'1')
- Time: O(R*C)
- Space: O(R*C) worst-case recursion/queue
"""
from typing import List


def num_islands(grid: List[List[str]]) -> int:
    if not grid or not grid[0]:
        return 0
    R, C = len(grid), len(grid[0])

    def dfs(r: int, c: int) -> None:
        if r < 0 or r >= R or c < 0 or c >= C or grid[r][c] != '1':
            return
        grid[r][c] = '0'
        dfs(r+1,c); dfs(r-1,c); dfs(r,c+1); dfs(r,c-1)

    count = 0
    for r in range(R):
        for c in range(C):
            if grid[r][c] == '1':
                count += 1
                dfs(r, c)
    return count


if __name__ == "__main__":
    g = [list("11000"), list("11000"), list("00100"), list("00011")]
    print(num_islands(g))  # 3

File: algorithms-practice/graphs/number_of_islands.cpp
// Number of Islands (DFS on grid)
// Time: O(R*C)
// Space: O(R*C)
#include <bits/stdc++.h>
using namespace std;

void dfs(vector<string>& g, int r, int c){
    int R = g.size(), C = g[0].size();
    if (r<0||r>=R||c<0||c>=C||g[r][c] != '1') return;
    g[r][c] = '0';
    dfs(g,r+1,c); dfs(g,r-1,c); dfs(g,r,c+1); dfs(g,r,c-1);
}

int numIslands(vector<string> g){
    int R = g.size(); if (!R) return 0; int C = g[0].size();
    int cnt = 0;
    for (int r=0;r<R;++r){
        for (int c=0;c<C;++c){
            if (g[r][c]=='1'){
                ++cnt; dfs(g,r,c);
            }
        }
    }
    return cnt;
}

int main(){
    vector<string> g = {"11000","11000","00100","00011"};
    cout << numIslands(g) << "\n"; // 3
    return 0;
}
